components/backup/src/softlimit.rs

```diff
@@ -1,97 +1,72 @@
 // Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
 
-use std::cell::RefCell;
 use std::cmp::Ordering as CmpOrder;
+
 use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::Arc;
 
 use super::Result;
 use collections::HashMap;
-use crossbeam::channel::{Receiver, Sender};
 
-use tikv_util::error;
 use tikv_util::metrics::ThreadInfoStatistics;
 use tikv_util::sys::SysQuota;
+use tokio::sync::{Semaphore, SemaphorePermit};
 
 /// SoftLimit is an simple "worker pool" just for
 /// restricting the number of workers can running concurrently.
-/// It is just a synchronous version of [tokio::sync::Semaphore].
-/// (With a poor(has O(n) space complex) but simple implementation.(not involved lock conditions))
-/// We should replace it with [tokio::sync::Semaphore] once backup has been refactored to asynchronous.
-#[derive(Clone)]
-pub struct SoftLimit {
-    sx: Sender<()>,
-    rx: Receiver<()>,
-    cap: Arc<AtomicUsize>,
+/// It is simply a wrapper over [tokio::sync::Semaphore],
+/// with a counter recording the current permits already and would grant.
+struct SoftLimitInner {
+    semaphore: Semaphore,
+    cap: AtomicUsize,
 }
 
-pub struct Guard {
-    sx: Sender<()>,
-}
-
-impl Drop for Guard {
-    fn drop(&mut self) {
-        if let Err(err) = self.sx.send(()) {
-            error!("failed to drop soft limit guard."; "err" => %err);
-        }
-    }
-}
+#[derive(Clone)]
+pub struct SoftLimit(Arc<SoftLimitInner>);
 
 impl SoftLimit {
-    fn enter(&self) -> Result<()> {
-        self.rx.recv()?;
-        Ok(())
-    }
-
     /// Makes a guard for one concurrent executing task.
     /// drop the guard for releasing the resource.
-    pub fn guard(&self) -> Result<Guard> {
-        self.enter()?;
-        Ok(Guard {
-            sx: self.sx.clone(),
-        })
+    pub async fn guard(&self) -> Result<SemaphorePermit<'_>> {
+        Ok(self.0.semaphore.acquire().await?)
     }
 
-    fn take_tokens(&self, n: usize) -> Result<()> {
-        for _ in 0..n {
-            self.rx.recv()?;
-        }
+    async fn take_tokens(&self, n: usize) -> Result<()> {
+        self.0.semaphore.acquire_many(n as _).await?.forget();
         Ok(())
     }
 
-    fn grant_tokens(&self, n: usize) -> Result<()> {
-        for _ in 0..n {
-            self.sx.send(())?;
-        }
+    async fn grant_tokens(&self, n: usize) -> Result<()> {
```

Comment:
```suggestion
    async fn grant_tokens(&self, n: usize) {
```

this function never errors now.